Remove support for OTP1/2 in parallel

miles-grant-ibigroup opened the pull request:
This PR removes the buggy support for making OTP1 and OTP2 requests in parallel.

In its place this PR provides full support for both OTP1 and OTP2 instances on their own.

V2 instances are now noted with adding a `v2: true` field in the api block. `api_v2` is no longer supported.

Closes https://github.com/opentripplanner/otp-react-redux/issues/563

Blocked by https://github.com/opentripplanner/otp-react-redux/pull/548, https://github.com/opentripplanner/otp-react-redux/pull/552

This PR actually wound up touching quite a lot of files.
### Things I'm looking for feedback on:
- As the primary effort here was to get things stable and to achieve feature parity, some architecture decisions are a little unclean. If you have good ideas on how to re-structure some of the api methods, these suggestions will be welcomed! However, please do keep in mind that the long-term goal is to move the entire api module to something far more modular like Apollo, removing the need for any of these methods
- A lot of work was done to match OTP1 response types as much as possible. If you have any ideas on how to remove the explicit re-writing and instead change downstream components to be more lenient in a way that is shorter than the current re-writing code, I would love that!
- OTP2 is missing some nice-to-haves that OTP1 currently enjoys, but I think the two are paired up enough for now (we can make the touch-ups in the future)
- Unrelated to this PR, but `STOPS_IN_BBOX` is firing a **lot**. I was not able to find an easy fix to include here

binh-dam-ibigroup commented on lib/actions/apiV1.js:
```diff
@@ -0,0 +1,233 @@
+  async function (dispatch, getState) {
+    await dispatch(findStop({ stopId: stop.stopId }))
+    const state = getState()
+    const { nearbyRadius: radius } = state.otp.config.stopViewer
```
Move to just above `if (radius ...)`. What is the fallback if `stopViewer` is not provided in the config?

miles-grant-ibigroup replied:
Copy pasted this from the old `api.js` but good catch with the `stopViewer` being undefined! I think all methods called in this action have radius defaults

miles-grant-ibigroup replied:
Refactored this a bit in d2507aa982ecca23a82fb1964c5e7a7099b81d88

binh-dam-ibigroup commented on lib/actions/apiV1.js:
```diff
@@ -0,0 +1,233 @@
+  }
+}
+
+export function findStop(params) {
+  return createQueryAction(
```
I kinda like the `const ... => createQueryAction(...)` notation (multiple instances).

miles-grant-ibigroup replied:
Me too!

miles-grant-ibigroup replied:
Refactored in d2507aa982ecca23a82fb1964c5e7a7099b81d88

binh-dam-ibigroup commented on lib/actions/apiV2.js:
```diff
@@ -0,0 +1,533 @@
+    }
+  )
+
+export function vehicleRentalQuery(
```
I kinda like the notation `const vehicleRentalQuery... => createGraphQLQueryAction(...)` (multiple instances).

miles-grant-ibigroup replied:
Refactored in d2507aa982ecca23a82fb1964c5e7a7099b81d88

binh-dam-ibigroup commented on lib/actions/apiV2.js:
```diff
@@ -0,0 +1,533 @@
+                stops: patternStops
+              }
+            })
+            route.patterns = routePatterns
+            // TODO: avoid explicit behavior shift like this
+            route.v2 = true
+
+            return route
```
In that case it's better to create a shallow copy of `payload.data` that gets modified (like in the other methods above) or, perhaps more preferably, perform a `clone`.

miles-grant-ibigroup replied:
Good call thanks! Addressed in d2507aa982ecca23a82fb1964c5e7a7099b81d88

binh-dam-ibigroup reviewed (changes requested):
Thanks for this PR, it is much more pleasant to read! I still need to play around more, but the one big change I need is to compute/initialize the correct version of the `actions` variable either using a function and maybe also using a redux state.

binh-dam-ibigroup commented:
API-wise, things seem to work!

miles-grant-ibigroup commented on lib/actions/form.js:
```diff
@@ -93,18 +95,22 @@ export function setQueryParam (payload, searchId) {
+export function parseUrlQueryString(params = getUrlParams(), source) {
   return function (dispatch, getState) {
+    const state = getState()
+    if (state.otp.ui.mainPanelContent !== null) return state
```
Our void returns were causing crashes! It turns out redux requires that each action returns the state object even if no changes are made.

miles-grant-ibigroup commented on lib/actions/ui.js:
```diff
@@ -172,7 +173,12 @@ export function matchContentToUrl(location) {
+          // This is a bit of a hack to check if the route details have been grabbed
+          // agencyName will only be populated if route details are repsent
+          // Moving away from manual requests should help resolve this
+          if (!state.otp.transitIndex?.routes?.[id]?.agencyName) {
```
Good call! Will use `bikesAllowed` instead which if unset in gtfs gets returned as `0`

miles-grant-ibigroup commented on lib/reducers/create-otp-reducer.js:
```diff
@@ -659,7 +659,7 @@ function createOtpReducer(config) {
         })
       case 'SET_VIEWED_STOP':
-        if (action.payload) {
+        if (action?.payload !== null) {
```
This action is sometimes called with `undefined` as the value. The explicit check avoids crashes

miles-grant-ibigroup commented on lib/reducers/create-otp-reducer.js:
```diff
@@ -715,18 +715,22 @@ function createOtpReducer(config) {
           transitIndex: {
             routes: {
-              [action.payload.routeId]: {
+              [action.payload.routeId || 'null']: {
```
Good call! Added

miles-grant-ibigroup commented on lib/actions/api.js:
```diff
@@ -15,17 +14,20 @@ import { convertToPlace, getPersistenceMode } from '../util/user'
   coreUtils.query
 const { randId } = coreUtils.storage
 
+// set actions variable that can be set to v2Actions if v2 server is set
+let actions = v1Actions
```
That's a good way to approach it thanks! Agreed it's clean for now and we can re-evaluate in the future.

miles-grant-ibigroup replied:
Done in e0bc80dc3dcc9b84645df4d5399b03c8de1788c9!

miles-grant-ibigroup commented on lib/actions/api.js:
```diff
@@ -245,16 +247,14 @@ function getJsonAndCheckResponse(res) {
 
-  const { api: apiV1, api_v2: apiV2, persistence } = state.otp.config
+  const { api, persistence } = state.otp.config
+  if (api.v2) actions = v2Actions
```
Refactored in e0bc80dc3dcc9b84645df4d5399b03c8de1788c9

miles-grant-ibigroup commented on lib/actions/api.js:
```diff
@@ -768,241 +530,35 @@ export function findRoutes(params) {
+    // Because findRoute is sometimes called from strange places,
+    // we must explicitly check for V2 as we can not rely on
+    // the check having already happened
+    if (api.v2) actions = v2Actions
```
Refactored in e0bc80dc3dcc9b84645df4d5399b03c8de1788c9

miles-grant-ibigroup commented on lib/actions/api.js:
```diff
@@ -1015,30 +571,21 @@ const findGeometryForPatternError = createAction(
-          })
+  return createQueryAction(
+    `index/patterns/${params.patternId}/geometry`,
+    findGeometryForPatternResponse,
+    findGeometryForPatternError,
```
I agree this could use work, but I think our energy here should be more focused on a full graphQL implementation can remove the need for any of these actions at all (and rather be one single action that can request any variable needed)

miles-grant-ibigroup commented:
Thanks for the suggestions!

binh-dam-ibigroup commented on lib/actions/api.js:
```diff
@@ -91,6 +88,24 @@ function getActiveItinerary(state) {
+ * @param {*} methodName    the method to execute
+ * @param  {...any} params  varargs of params to send to the action
+ */
+function executeOTPVersionSpecificAction(methodName, ...params) {
```
Pretty cool! Maybe you can rename to just `executeOTPAction`.

miles-grant-ibigroup replied:
We sure can! Good suggestion thank you

miles-grant-ibigroup replied:
Renamed in a500cbe779f68c61c9e9739c30b640641fdb2b1a

binh-dam-ibigroup commented on lib/actions/apiV2.js:
```diff
@@ -496,20 +503,19 @@ export function findRoute(params) {
             })
-            route.patterns = routePatterns
+            newRoute.patterns = routePatterns
             // TODO: avoid explicit behavior shift like this
```
Can this comment be removed now?

miles-grant-ibigroup replied:
I wish... the problem for me is the `v2` field which I don't like. I think ideally the rest of OTP shouldn't know if it's dealing with OTP1 or OTP2

binh-dam-ibigroup commented on lib/actions/ui.js:
```diff
@@ -174,9 +174,9 @@ export function matchContentToUrl(location) {
         if (id) {
           // This is a bit of a hack to check if the route details have been grabbed
-          // agencyName will only be populated if route details are repsent
+          // bikesAllowed will only be populated if route details are repsent
```
Typo "~repsent~" => "present"

binh-dam-ibigroup replied:
Not sure about `bikesAllowed` either - that field is not populated by many Atlanta agencies.

binh-dam-ibigroup replied:
Would the presence of at least one pattern be a more reliable criterion?

miles-grant-ibigroup replied:
Good catch thank you! and `bikesAllowed` feels strong to me because even if it's not populated OTP still sets it to 0 which allows us to do this null check

miles-grant-ibigroup replied:
One pattern was what I initially had, but ran into problems where sometimes routes wouldn't have patterns (?) although the full route was downloaded.

miles-grant-ibigroup replied:
Typo fixed in a500cbe779f68c61c9e9739c30b640641fdb2b1a

binh-dam-ibigroup reviewed (commented):
The refactor is nice! I am wondering about using `bikesAllowed` to detect downloading route details.

miles-grant-ibigroup commented:
Thanks Binh, changes are addressed

demory commented:
Have you seen this error when loading the stop viewer (either via an itinerary or a map popup) when running against OTP2? I'm seeing it consistently. Works fine w/ OTP1.

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/653100/158480698-8ae86018-1190-4a2c-81cd-13d509354203.png">

miles-grant-ibigroup commented:
Very good catch thank you! That's a regression that came in when I moved the nearby endpoint to OTP2. Fixed in e8ff07d4c60bdb564eeb79b4e4fb28486c1da907

demory commented:
Thanks Miles. I've been testing w/ the latest changes and now am seeing a couple other OTP2-specific UI issues.

1. In the stop viewer, any departure time for the "related stops" always shows as 12:00am:

<img width="426" alt="image" src="https://user-images.githubusercontent.com/653100/158847113-2c0b17c5-5c04-4ec6-8579-5a2b6c55b47d.png">

2. Navigating from the trip viewer to the stop viewer using one of the stop-specific "View" buttons causes a crash:

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/653100/158847450-5c1c87d5-cf2a-4c9e-b0e7-6e63418c0254.png">

(Also, I'm getting some odd OTP2 results using the ATL Rides instance, e.g. no transit trips when I definitely expect them. But that could be an OTP issue, need to investigate further.)

miles-grant-ibigroup commented:
Thanks @demory for spotting those regressions! Things should be back to stability in 24adf0b688525050b8c29e7cd68a1d1248698679

miles-grant-ibigroup commented:
@demory midnight bus departures are resolved in 1633f44c8867253ba13bd0d9c0aed65a4043da92, the stops layer not working error was actually not a client-side issue but a server side issue and one that I've fixed [here](https://github.com/ibi-group/OpenTripPlanner/pull/99). I do still think moving to graphQL for those requests is the way to go, but that can be a separate PR. I'm also taking a look at the snapping, but will push any fixes for that to a new PR. 

Please let me know if there are any other blocking changes!
